File: programmers/etc/level3/최적의_행렬_곱셈.py
import sys
INF = sys.maxsize

def solution(m):
    n = len(m)
    dp = [[INF]*n for _ in range(n)]
    for i in range(n):
        dp[i][i] = 0
    
    for gap in range(1, n):
        for start in range(n):
            end = start+gap
            if end >= n: break
            for sep in range(start, end):
                value = dp[start][sep]+dp[sep+1][end]+(m[start][0]*m[sep][1]*m[end][1])
                dp[start][end] = min(dp[start][end], value)
    
    return dp[0][-1]


m = [[5,3],[3,10],[10,6]]
print(solution(m)) # 270

File: Java/programmers/etc/level3/최적의행렬곱셈.java
package CodeTest.Java.programmers.etc.level3;

import java.util.*;

public class 최적의행렬곱셈 {
    public int solution(int[][] m) {
        int n = m.length;
        int[][] dp = new int[n][n];
        
        for (int[] d : dp) {
            Arrays.fill(d, Integer.MAX_VALUE);
        }
        
        for (int i=0; i<n; i++) {
            dp[i][i] = 0;
        }
        
        for (int gap=1; gap<n; gap++) {
            for (int start=0; start<n; start++) {
                int end = start+gap;
                if (end >= n) break;
                for (int sep=start; sep<end; sep++) {
                    int value = dp[start][sep]+dp[sep+1][end]+(m[start][0]*m[sep][1]*m[end][1]);
                    dp[start][end] = Math.min(dp[start][end], value);
                }
            }
        }
        
        return dp[0][n-1];
    }
}
